core/src/qrcode/MQRDataBlock.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2016 Nu-book Inc.
+ * Copyright 2016 ZXing authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MQRDataBlock.h"
+
+#include "MQRVersion.h"
+#include "QRErrorCorrectionLevel.h"
+#include "ZXContainerAlgorithms.h"
+
+namespace ZXing::MicroQRCode {
+
+std::vector<DataBlock> DataBlock::GetDataBlocks(const ByteArray& rawCodewords, const Version& version,
+												QRCode::ErrorCorrectionLevel ecLevel)
+{
+	if (Size(rawCodewords) != version.totalCodewords())
+		return {};
+
+	// Figure out the number and size of data blocks used by this version and
+	// error correction level
+	auto& ecBlocks = version.ecBlocksForLevel(ecLevel);
+
+	// First count the total number of data blocks
+	int totalBlocks = ecBlocks.numBlocks();
+	if (totalBlocks == 0)
+		return {};
```

Comment:
The reason for the duplication is because there are two Version classes - `QRVersion` and `MQRVersion` - which means that there needs to be two `GetDataBlock` functions.  What I'll do is change the construction functions `ProvisionalVersionForDimension,` `VersionForNumber,` `DecodeVersionInformation` to have a bool flag to indicate if a micro QR code is being processed.  If true, I can return a micro QR code version of QRVersion for use in the rest of the decoding process.  This should simplify the rest of the decoding code.

🤞 my IDE doesn't try to reformat all the code using clang-format as I make these changes 😄

Reply:
Perfect. That sounds like what I had in mind in my comment from yesterday. About the clang-format issue: removing the file while working on this might do the trick? Or what I use extensively during commit is `git gui` and its ability to add individual hunks or even lines to the staging area. Thanks for the effort!